src/common.rs

```diff
@@ -165,8 +171,52 @@ fn zip_apply<'a, P: Iterator<Item = Option<JsonPath<'a>>>, C: FromIterator<Optio
     object_lookup: bool,
 ) -> DataFusionResult<ArrayRef> {
     if let Some(d) = json_array.as_any_dictionary_opt() {
-        let a = zip_apply(d.values(), path_array, to_array, jiter_find, object_lookup)?;
-        return Ok(d.with_values(a).into());
+        let values = zip_apply(d.values(), path_array, to_array, jiter_find, object_lookup)?;
+        if !is_json_union(values.data_type()) {
+            return Ok(d.with_values(values));
+        }
+        // JSON union: post-process the array to set keys to null where the union member is null
+        let type_ids = values.as_any().downcast_ref::<UnionArray>().unwrap().type_ids();
+
+        // FIXME downcast_dictionary_array! not hygenic with datafusion import
+        let masked = match d.keys().data_type() {
+            // FIXME downcast_integer! not hygenic with datafusion import
+            datafusion::arrow::datatypes::DataType::Int8 => Arc::new(DictionaryArray::new(
```

Comment:
```suggestion
            DataType::Int8 => Arc::new(DictionaryArray::new(
```
and below, might make this easier to read?